src/engine/joinOrdering/IKKBZ.h

```diff
@@ -0,0 +1,119 @@
+// Copyright 2024, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author:
+// Mahmoud Khalaf (2024-, khalaf@cs.uni-freiburg.de)
+
+#pragma once
+
```

Comment:
Okay, to improve this design I would suggest the following first step:

1. Get the combining of nodes right using the follwing design:

```
class Relation {
 // ID, selectivity, cardinality
};

// created by the NormalizeStep.
class CombinedRelations {
  std::vector<Relation>; // contained relations.
  double rank() ; // compute the rank.
}

// Called for two consecutive elements in a chain where the rank is out of order
CombinedRelations combine(const CombinedRelations& first, const CombinedRelations& second);


// Run IKKBZ-Normalize on a single chain until it is fully normalized.
void IkkbzNormalize(std::vector<CombinedRelations>& chain);

// Givien the above structure, this is now simple, just merge by rank.
std::vector<CombinedRelations> IkkbzMerge(std::vector<std::vector<CombinedRelations>>);
```


Maybe start this design in a separate PR, and then star with unit tests for those rather small functions.

Then in a much simpler IKKBZ graph, you can incorporate those structures, but maybe let me know first, 
so that we can talk about the next steps.

Reply:
I think this is the thread that is worth pursuing for now so we can build up a correct and maintainable implementation of IKKBZ.